src/evict/evict_lru.c

```diff
@@ -843,6 +835,164 @@ __wt_evict_file_exclusive_off(WT_SESSION_IMPL *session)
+	delta_pages = pgs_evicted_cur - conn->evict_tune_pgs_last;
+	pgs_evicted_persec_cur = (delta_pages * WT_THOUSAND) / delta_millis;
+
+	/*
+	 * If we have not recorded the pages evicted per second rate in the
+	 * previous interval, we need to wait for another measurement interval
+	 * until we can record that rate, because we cannot make any tuning
+	 * decisions without it.
+	 */
+	if (conn->evict_tune_pg_sec_last == 0)
+		goto err;
+
+	pct_diff = (int)(pgs_evicted_persec_cur - conn->evict_tune_pg_sec_last)
+	    * 100 / (int) conn->evict_tune_pg_sec_last;
+
+	/*
+	 * If we previously added threads and saw no improvement, we will remove
+	 * half of the threads currently working.
+	 * If we added threads and saw improvement, we will keep adding.
+	 * If we previously removed threads and so saw no degradation, we will
+	 * keep removing.
+	 * If we previously removed threads and saw degradation, we will add.
+	 */
+	if (conn->evict_tune_last_action == EVICT_ADDED)
+		if (pct_diff > EVICT_TUNE_THRESHOLD)
+			try_create_threads = true;
+		else
+			try_remove_threads = true;
+	else if (conn->evict_tune_last_action == EVICT_REMOVED)
+		if (pct_diff < -EVICT_TUNE_THRESHOLD)
+			try_create_threads = true;
+		else
+			try_remove_threads = true;
+	else /* We never created or removed threads. Add some */
+		try_create_threads = true;
+
+	WT_ASSERT(session, try_create_threads ^ try_remove_threads);
+
+	if (try_create_threads) {
+		cur_threads = conn->evict_threads.current_threads;
+		target_threads = (conn->evict_threads.max - cur_threads) / 2;
+		if (target_threads == 0)
+			target_threads = 1;
+
+		for (i = 0; i < target_threads; i++) {
+			/*
+			 * Attempt to create a new thread if we have capacity
+			 * and if the eviction goals are not met.
+			 */
+			if (F_ISSET(cache, WT_CACHE_EVICT_ALL))
+				WT_ERR(__wt_thread_group_start_one(
+					       session, &conn->evict_threads,
+					       false));
+
+			if (conn->evict_threads.current_threads == cur_threads)
+				break;
+
+			/* We created a new thread. Keep track of it. */
+			cur_threads++;
+			conn->evict_tune_last_action = EVICT_ADDED;
+			WT_STAT_CONN_INCR(session,
+					  cache_eviction_worker_created);
+			WT_STAT_CONN_SET(session,
+					 cache_eviction_active_workers,
+					 conn->evict_threads.current_threads);
+			__wt_verbose(session, WT_VERB_EVICTSERVER,
+				     "added thread");
+		}
+	}
+	else if (try_remove_threads) {
+		cur_threads = conn->evict_threads.current_threads;
+		target_threads = (cur_threads - conn->evict_threads.min) / 2;
+
+		for (i = 0; i < target_threads; i++) {
+			WT_ERR(__wt_thread_group_stop_one(
+				       session, &conn->evict_threads, false));
+
+			if (conn->evict_threads.current_threads == cur_threads)
+				break;
+
+			conn->evict_tune_last_action = EVICT_REMOVED;
+			WT_STAT_CONN_INCR(session,
+			    cache_eviction_worker_removed);
+			WT_STAT_CONN_SET(session, cache_eviction_active_workers,
```

Comment:
Since a thread is always created or removed, maybe we can take this stat update and the one in the if loop outside the loops and just update once.